runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -0,0 +1,436 @@
+     * is actually performed.
+     *
+     * @param p java predicate (function to evaluate)
+     * @return a view of the values contained in this map meeting the predicate condition.
+     */
+    @Accessor
+    public List<V> scanAndFilter(Predicate<? super V> p) {
+        return mainMap.values().parallelStream().filter(p).collect(Collectors.toList());
+    }
+
+    /**
+     * Returns a {@link Collection} filtered by entries (keys and/or values).
+     * This method has a memory/CPU advantage over the map iterators as no deep copy
+     * is actually performed.
+     *
+     * @param entryPredicate java predicate (function to evaluate)
+     * @return a view of the entries contained in this map meeting the predicate condition.
+     */
+    @Accessor
+    public Collection<Map.Entry<K, V>> scanAndFilterByEntry(Predicate<? super Map.Entry<K, V>>
+                                                                    entryPredicate) {
+        return mainMap.entrySet().parallelStream().filter(entryPredicate).collect(Collectors
+                .toCollection(ArrayList::new));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @MutatorAccessor(name = "remove", undoFunction = "undoRemove",
+                                undoRecordFunction = "undoRemoveRecord")
+    @SuppressWarnings("unchecked")
+    public V remove(@ConflictParameter Object key) {
+        V previous =  mainMap.remove(key);
+        unmapSecondaryIndexes((K) key, previous);
+        return previous;
+    }
+
+    @DontInstrument
+    protected V undoRemoveRecord(CorfuTable<K, V, F, I> table, K key) {
+        return table.mainMap.get(key);
+    }
+
+    @DontInstrument
+    protected void undoRemove(CorfuTable<K, V, F, I> table, V undoRecord, K key) {
+        if (undoRecord == null) {
+            V previous =  table.mainMap.remove(key);
+            table.unmapSecondaryIndexes(key, previous);
+        } else {
+            V previous = table.mainMap.put(key, undoRecord);
+            if (table.indexFunctions.size() > 0) {
+                table.unmapSecondaryIndexes(key, previous);
+                table.mapSecondaryIndexes(key, undoRecord);
+            }
+        }
+    }
+
+    @Mutator(name = "remove", noUpcall = true)
+    void delete(@ConflictParameter K key) {
+        V previous =  mainMap.remove(key);
+        unmapSecondaryIndexes(key, previous);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Mutator(name = "putAll",
+            undoFunction = "undoPutAll",
+            undoRecordFunction = "undoPutAllRecord",
+            conflictParameterFunction = "putAllConflictFunction")
+    public void putAll(@Nonnull Map<? extends K, ? extends V> m) {
+        // If we have no index functions, then just directly put all
+        if (indexFunctions.isEmpty()) {
+            mainMap.putAll(m);
+        }
+        // Otherwise we must update all secondary indexes
+        // TODO: Do this in parallel (need to acquire update locks, potentially)
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)